controller/engine_image_controller.go

```diff
@@ -385,6 +384,49 @@ func (ic *EngineImageController) syncEngineImage(key string) (err error) {
 	return nil
 }
 
+func (ic *EngineImageController) syncNodeDeploymentMap(engineImage *longhorn.EngineImage) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "cannot sync NodeDeploymenMap for engine image %v", engineImage.Name)
+	}()
+
+	nodeDeploymentMap := map[string]bool{}
+
+	nodes, err := ic.ds.ListNodes()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		nodeDeploymentMap[node.Name] = false
+	}
+
+	// Get all pods of the DaemonSet corresponding to the engine image
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: types.GetEngineImageLabels(engineImage.Name),
+	})
+	if err != nil {
+		return err
+	}
+	EIDaemonSetPods, err := ic.ds.ListPodsBySelector(selector)
+	if err != nil {
+		return err
+	}
+	for _, pod := range EIDaemonSetPods {
+		allContainerReady := true
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			if !containerStatus.Ready {
+				allContainerReady = false
+				break
+			}
+		}
+		nodeDeploymentMap[pod.Spec.NodeName] = allContainerReady
+	}
+
+	if !reflect.DeepEqual(engineImage.Status.NodeDeploymentMap, nodeDeploymentMap) {
```

Comment:
This isn't really needed, since the actual deepEquals comparison existingEngineImage newEngineImage in the controller will take care of it. 

You might think this is a performance optimization, but since the nodeDeploymentMap is always a new map this deepEquals will need to traverse the whole map every time.

So you are not gaining anything, because if you leave it out you just let the controller deepEquals traverse it.
And in the case where the map has changed, you will end up traversing the map twice :)

Reply:
Great point! Thank you. Fixed